sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotHideBaseClassMethods.cs

```diff
@@ -0,0 +1,180 @@
+
+                    if (classSymbol == null || classDeclaration.Identifier.IsMissing)
+                    {
+                        return;
+                    }
+
+                    var issueFinder = new IssueFinder(classSymbol, c.SemanticModel);
+
+                    classDeclaration
+                        .Members
+                        .OfType<MemberDeclarationSyntax>()
+                        .Select(issueFinder.FindIssue)
+                        .WhereNotNull()
+                        .ToList()
+                        .ForEach(c.ReportDiagnostic);
+                },
+                SyntaxKind.ClassDeclaration);
+        }
+
+        private class IssueFinder
+        {
+            private readonly IList<IMethodSymbol> allBaseClassMethods;
+            private readonly SemanticModel semanticModel;
+
+            public IssueFinder(INamedTypeSymbol classTypeSymbol, SemanticModel semanticModel)
+            {
+                this.semanticModel = semanticModel;
+                allBaseClassMethods = GetAllBaseMethods(classTypeSymbol).ToList();
+            }
+
+            private static IEnumerable<IMethodSymbol> GetAllBaseMethods(INamedTypeSymbol classTypeSymbol)
+            {
+                while (classTypeSymbol?.BaseType != null)
+                {
+                    var members = classTypeSymbol.BaseType
+                            .GetMembers()
+                            .OfType<IMethodSymbol>()
+                            .Where(m => m.DeclaredAccessibility != Accessibility.Private)
+                            .Where(m => m.Parameters.Length > 0)
+                            .Where(m => !string.IsNullOrEmpty(m.Name));
+
+                    foreach (var m in members)
+                    {
+                        yield return m;
+                    }
+
+                    classTypeSymbol = classTypeSymbol.BaseType;
+                }
+            }
+
+            public Diagnostic FindIssue(MemberDeclarationSyntax memberDeclaration)
+            {
+                var methodSymbol = semanticModel.GetDeclaredSymbol(memberDeclaration) as IMethodSymbol;
+                var issueLocation = (memberDeclaration as MethodDeclarationSyntax)?.Identifier.GetLocation();
+
+                if (methodSymbol == null || issueLocation == null)
+                {
+                    return null;
+                }
+
+                var baseMethodHidden = FindBaseMethodHiddenByMethod(methodSymbol);
+                return baseMethodHidden != null ? Diagnostic.Create(rule, issueLocation, baseMethodHidden) : null;
+            }
+
+            private IMethodSymbol FindBaseMethodHiddenByMethod(IMethodSymbol methodSymbol)
+            {
+                var baseMemberCandidates = allBaseClassMethods.Where(m => m.Name == methodSymbol.Name);
+
+                IMethodSymbol hiddenBaseMethodCandidate = null;
+                var hasBaseMethodWithSameSignature = baseMemberCandidates.Any(baseMember =>
```

Comment:
Lambda seems like a good candidate to a method. Also I am not a big fan of a captured variable, I would use `FirstOrDefault` instead of `Any`.